File: Workshop2_Linked Lists/Problem3_Merge_Two_Sorted_Lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def mergeTwoLists(self, l1: ListNode, l2: ListNode) -> ListNode:
        dummy = ListNode(0) #create a dummy node to point to the head of the list.
        cur = dummy
        while l1 != None and l2 != None:    #while we have not run to the end of either list.
            if l1.val < l2.val: #l1's node comes before l2's node
                cur.next = l1
                l1 = l1.next
            else:               #l2's node comes before l1's node
                cur.next = l2
                l2 = l2.next
            cur = cur.next
        cur.next = l1 or l2     # the list that is not yet reached the end is appended to the end.

        return dummy.next       # we return dummy.next because it is the first node in the combined list.

File: Workshop2_Linked Lists/Problem3_Merge Two Sorted Lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
        if (l1 == NULL && l2 == NULL)
        {
            return NULL;
        }
        ListNode res(0); //dummy head 
        ListNode* newList = &res; //newList will have a dummy head 
        while(l1 != NULL && l2 != NULL){
            if(l1->val <= l2->val)
            {
                newList->next = l1; //append the according values to the next part of the list
                l1 = l1->next; //iterate 
            }
            else
            {
                newList->next = l2;
                l2 = l2->next; //iterate
            }
            newList = newList->next; //go to the end of the newList
        }
        //if we break out the while loop,either theres more l1 left, l2 left or none, plan accordingly.
        if (l1) 
        {
            newList->next = l1;
        }
        else if (l2)
        {
            newList->next = l2;
        }
        return res.next; //return everything after dummy head
    }
};